include/oneapi/dpl/pstl/utils.h

```diff
@@ -128,6 +128,21 @@ class __pstl_assign
     }
 };
 
+template <typename _Comp, typename _Proj>
+class __compare
+{
+    _Comp __comp;
+    _Proj __proj;
+
+  public:
+    template <typename _Xp, typename _Yp>
+    bool
+    operator()(_Xp&& __x, _Yp&& __y)
+    {
+        return __comp(__proj(std::forward<_Xp>(__x)), __proj(std::forward<_Yp>(__y)));
+    }
```

Comment:
Why not keep `const&` parameters as in the original lambda? I believe neither comparators nor projections are supposed to modify their arguments.

Reply:
good question. I intended keep const&, but made approach at least  consistency with the other similar predicates (__pstl_equal, __pstl_less, etc).
The next question why the other similar predicates accept T&&, not const T&... Probably, we had a compile issue with some tricky types (a kind proxy types etc). I will try to figure out it.

Reply:
Actually, I've remembered a case when we had a compilation issue when we had changed `T&&` by const `T&`. It was the binary_search.pass test. That code tries to call a lambda with T& argument, which is const T& in its turn.

So, keeping `const T&` for argument of `__compare` doesn't matche to operator `__proj`, which can be implemented by user as`operator()(T&) {}`. 

So, keeping `T&` we make oneDPL more robust relative to a user projection type definition.

Reply:
If I follow correctly, you are saying that using forwarding references here allows us to accommodate the case where a user provides a projection operator which takes as input a **non-const lvalue ref** (`T&`).  

The question then follows: is this a good thing?  

Our previous code did not enable this, it would be a compile time error.  If a user's projection operator actually modifies the input reference, I believe that would result in UB in our sort routine (modified inputs every time the projection is applied).  Conceptually projection operators shouldn't modify the input.  So, what we are enabling is "convenience" for the user of not adding a `const` to their input params, and possibly letting a compile time error become a runtime error if their projection actually does modify the input.

Reply:
C++17 states that predicates (unary and binary) as well as comparators "shall/will not apply any non-constant function through the dereferenced iterator". For parallel algorithms, it additionally states that unary and binary operations also "shall not directly or indirectly modify objects via their arguments".

In C++20, projections, comparators, predicates etc. for range-based algorithms are required to satisfy the `regular_invocable` concept. That concept states that the function invocation "shall not modify the
function object or the arguments."

That's what I meant above saying that "neither comparators nor projections are supposed to modify their arguments." Even if user-provided function objects can technically have non-constant call operators/parameters, its only benefit seems to allow users "forget" const qualifiers - at the risk of having runtime errors, like Dan said above.

Unfortunately, both std libraries and Thrust are tolerant to function objects omitting `const`, and this is the main reason why we may need to do the same.

Reply:
So, if we want to keep `T&&`, I think it should be consistency with the similar predicates in the same file, at least. I mean file `utils.h` and `__pstl_equal`, `__pstl_less`, `__pstl_greater` etc.

Reply:
One one hand, in "narrow sight" I understand: We had a customer issue with  "device copyable", found a reason as lambda-comparator with the captured user comparator projection and "`const T&` call convention". Looking just only that, no considering the situation in general, yes we can just keep `const T&`, as we had in the mentioned "lambda" version. It can be one commit (one PR).
On the other hand,
1) I don't like to make "the inconsistency" bigger  and to "seed" a potential oneDPL issue when a user  will pass "the comparators / projection operators to allow non-const qualified lvalue ref parameters". I understand that it was before, with lambda,  but any any case - if we change that place  - lets looking in the future.
2) Moreover, having that inconsistency in the similar predicates in this file confuses the others oneDPL contributors. Somebody might add a new predicate with `T&` someone  - with `const T&`. The mess will be increased.
3) Also that inconsistency will confuse the reviewers when we merge it to LLVM upstream. 

As an option, that I would propose  - 1) to merge the "device copyable" fix for `__parallel_stable_sort` with `const T&`, as it was in lambda. 2) right after that - the next change: to eliminate the mentioned inconsistency for the all predicates, at least "comparing" type.

Reply:
I'm OK with this plan.  I'll note that if we make this change and release, it becomes much more difficult to revert this change since it could then "break" working user code. 

Do you intend to make changes for more alignment, or do you think all that is required is contained in the existing PR?

Reply:
1) done
2) It seems we tend to relax the requirement  - to use `T&&` instead for `const T&` for input parameter - so, it doesn't  break working user code.

Reply:
I think you misunderstood what I'm saying.  
I agree this PR does not break working user code.  The purpose of my comment is to state that it is "easy" to go in this direction.  However, if we later were to change our minds and want to go back to `const T&`, from `T&&`, that could break working user code, so we would not make that change. 
I'm just mentioning it to make it clear that we should be sure about our decision because it would be difficult to revert.

Reply:
I see your point, Dan....
Alexey (@akukanov),  It seems we need a decision maker here.

Reply:
One more argument in favor of `T&&` - std::invoke accepts arguments as `Args&&... args`

Reply:
I am OK with the current patch. As for the plan - I do not object to revisiting the content of `pstl/utils/h` but I do not think it should be done for the same milestone.